src/shared/Microsoft.Git.CredentialManager/Settings.cs

```diff
@@ -0,0 +1,176 @@
+        /// </summary>
+        /// <param name="repositoryPath">Optional path of a repository to lookup local configuration from.</param>
+        /// <param name="remoteUri">Optional git remote address that settings should be scoped to.</param>
+        /// <param name="envarName">Optional environment variable name.</param>
+        /// <param name="section">Optional Git configuration section name.</param>
+        /// <param name="property">Git configuration property name. Required if <paramref name="section"/> is set, optional otherwise.</param>
+        /// <param name="value">Value of the requested setting.</param>
+        /// <returns>True if a setting value was found, false otherwise.</returns>
+        public bool TryGetSetting(string repositoryPath, Uri remoteUri, string envarName, string section, string property, out string value)
+        {
+            IEnumerable<string> allValues = GetSettingValues(repositoryPath, remoteUri, envarName, section, property);
+
+            value = allValues.FirstOrDefault();
+
+            return value != null;
+        }
+
+        /// <summary>
+        /// Try and get the all values of a specified setting as specified in the environment and Git configuration,
+        /// in the correct order or precedence.
+        /// </summary>
+        /// <param name="repositoryPath">Optional path of a repository to lookup local configuration from.</param>
+        /// <param name="remoteUri">Optional git remote address that settings should be scoped to.</param>
+        /// <param name="envarName">Optional environment variable name.</param>
+        /// <param name="section">Optional Git configuration section name.</param>
+        /// <param name="property">Git configuration property name. Required if <paramref name="section"/> is set, optional otherwise.</param>
+        /// <returns>All values for the specified setting, in order of precedence, or an empty collection if no such values are set.</returns>
+        public IEnumerable<string> GetSettingValues(string repositoryPath, Uri remoteUri, string envarName, string section, string property)
+        {
+            string value;
+
+            if (envarName != null)
+            {
+                if (_environment.TryGetValue(envarName, out value))
+                {
+                    yield return value;
+                }
+            }
+
+            if (section != null && property != null)
+            {
+                using (var config = _git.GetConfiguration(repositoryPath))
+                {
+                    if (remoteUri != null)
+                    {
+                        /*
+                         * Look for URL scoped "section" configuration entries, starting from the most specific
+                         * down to the least specific (stopping before the TLD).
+                         *
+                         * In a divergence from standard Git configuration rules, we also consider matching URL scopes
```

Comment:
@jrbriggs, calling out this part of the code again in the latest push.
I realised that GCM Core was doing something 'not consistent with Git', as does GCM Windows.

Git really only expects `[credential]` configuration entries to have URLs that _include the scheme_ such as:

```ini
[credential "https://dev.azure.com"]
    useHttpPath = true
    proxy = https://user:pass@proxy.example.com
```

..but the initial code was only looking for entries without the scheme.. i.e,

```ini
[credential "dev.azure.com"]
    useHttpPath = true
    proxy = https://user:pass@proxy.example.com
```

GCM Windows seems to have introduced the idea of omitting the scheme, probably so that you can specify options for multiple protocols at the same time, such as "http" and "https".

I have adjusted (and called out the divergence) the code here to look for both the scheme- and scheme-less- variants so that we can be consistent with Git's logic first, and then also pick up the behaviour of GCM Windows second.

---
**Side note:**
Looks like GCM Windows introduced the `credential.proxy` configuration entry and the `GCM_HTTP_PROXY` environment variable. Git/cURL have their own configuration of `http.proxy` and `HTTP_PROXY`. GCM Windows seems to support `GCM_HTTP_PROXY` but emit a deprecation warning message.

This is something I'll need to keep in mind in the next PR when reading proxy config, and trying to stay compat with existing GCM Windows users' setups.

Reply:
@mjcheetham thanks for calling that out. I think having back-compat makes sense. Should we add a unit test for `GetGitConfigurationScopes` with scheme-less URIs?

Reply:
> Should we add a unit test for `GetGitConfigurationScopes` with scheme-less URIs?

Do you mean a test like `GetGitConfigurationScopes(new Uri("no-scheme.com"))`?
A `System.Uri` _must_ have a scheme.

If you mean a test of the `Settings.TryGetSetting` and `Settings.GetSettingValues` methods then that already exists with test `Settings_GetSettingValues_EnvarAndMultipleConfig_ReturnsAllWithCorrectPrecedence`.

We create a Git configuration with entries containing no scope, a scope with a scheme, and a scope without a scheme (and an environment variable).

```csharp
public void Settings_GetSettingValues_EnvarAndMultipleConfig_ReturnsAllWithCorrectPrecedence()
{
    const string repositoryPath = "/tmp/repos/foo/.git";
    const string remoteUrl = "http://example.com/foo.git";
    const string scope1 = "http://example.com";
    const string scope2 = "example.com";
    const string envarName = "GCM_TESTVAR";
    const string section = "gcmtest";
    const string property = "bar";
    var remoteUri = new Uri(remoteUrl);

    const string value1 = "First value";
    const string value2 = "Second value";
    const string value3 = "Third value";
    const string value4 = "Last value";

    string[] expectedValues = {value1, value2, value3, value4};

    var envars = new EnvironmentVariables(new Dictionary<string, string>
    {
        [envarName] = value1,
    });
    var git = new TestGit();
    git.AddRepository(repositoryPath, new Dictionary<string, string>
    {
        [$"{section}.{scope1}.{property}"] = value2,
        [$"{section}.{scope2}.{property}"] = value3,
        [$"{section}.{property}"]          = value4
    });

    var settings = new Settings(envars, git);
    string[] actualValues = settings.GetSettingValues(repositoryPath, remoteUri, envarName, section, property).ToArray();

    Assert.Equal(expectedValues, actualValues);
}
```